staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/collections.go

```diff
@@ -0,0 +1,221 @@
+	}
+	if listType.Field(0).Tag.Get("protobuf") != "" {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf(`expected TypeMeta protobuf field tag to be ""`)
+	}
+	// ListMeta
+	listMeta, ok := listValue.Field(1).Interface().(metav1.ListMeta)
+	if !ok {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf("expected ListMeta field to have ListMeta type")
+	}
+	if listType.Field(1).Tag.Get("protobuf") != "bytes,1,opt,name=metadata" {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf(`expected ListMeta protobuf field tag to be "bytes,1,opt,name=metadata"`)
+	}
+	// Items
+	items, err := meta.ExtractList(list)
+	if err != nil {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, err
+	}
+	if listType.Field(2).Tag.Get("protobuf") != "bytes,2,rep,name=items" {
+		return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf(`expected Items protobuf field tag to be "bytes,2,rep,name=items"`)
+	}
+	for _, item := range items {
+		if _, ok := item.(proto.Sizer); !ok {
+			return metav1.TypeMeta{}, metav1.ListMeta{}, nil, fmt.Errorf(`expected Items elements to implement proto.Sizer`)
+		}
+	}
+	return typeMeta, listMeta, items, nil
+}
+
+func streamingEncodeUnknownList(w io.Writer, unk runtime.Unknown, listMeta metav1.ListMeta, items []runtime.Object) error {
+	_, err := w.Write(protoEncodingPrefix)
+	if err != nil {
+		return err
+	}
+	// TypeMeta
+	_, err = encodeValue(w, []byte{0xa}, &unk.TypeMeta)
+	if err != nil {
+		return err
+	}
+	// Raw
+	_, err = w.Write([]byte{0x12})
+	if err != nil {
+		return err
+	}
+	err = streamingEncodeList(w, listMeta, items)
+	if err != nil {
+		return err
+	}
+	// ContentEncoding
+	_, err = encodeValue(w, []byte{0x1a}, unk.ContentEncoding)
+	if err != nil {
+		return err
+	}
+	// ContentType
+	_, err = encodeValue(w, []byte{0x22}, unk.ContentType)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func streamingEncodeList(w io.Writer, listMeta metav1.ListMeta, items []runtime.Object) error {
+	size, err := itemsSize(items)
+	if err != nil {
+		return err
+	}
+	size += embededSize(&listMeta)
+	// ListMeta
+	_, err = writeVarintGenerated(w, size)
+	if err != nil {
+		return err
+	}
+	listMetaSize, err := encodeValue(w, []byte{0xa}, &listMeta)
+	if err != nil {
+		return err
+	}
+	// Items
+	listSize, err := streamingEncodeObjectSlice(w, items)
+	if err != nil {
+		return err
+	}
+	if listMetaSize+listSize != size {
+		return fmt.Errorf("the size value was %d, but encoding wrote %d bytes to data", size, listMetaSize+listSize)
+	}
+	return nil
+}
+
+func streamingEncodeObjectSlice(w io.Writer, items []runtime.Object) (size int, err error) {
+	for _, item := range items {
+		n, err := encodeValue(w, []byte{0x12}, item)
+		size += n
+		if err != nil {
+			return size, err
+		}
+	}
+	return size, nil
+}
+
+func encodeValue(w io.Writer, field []byte, value any) (size int, err error) {
```

Comment:
I think we're calling this 5 times currently:
* runtime.Unknown typeMeta / contentEncoding / contentType - can we switch to calling something local to runtime.Unknown that knows how to encode these things?
* individual runtime.Objects - can we switch to something like RawSerializer#doEncode that is more robust?
* listMeta - I think this remains the only thing we still need to encode in detail here, so can we just inline this of making a generic method?

Reply:
> runtime.Unknown typeMeta / contentEncoding / contentType - can we switch to calling something local to runtime.Unknown that knows how to encode these things?

There are no local functions as all code is autogenerated into one function.

Reply:
> individual runtime.Objects - can we switch to something like RawSerializer#doEncode that is more robust?

I'm worried about how many additional edge cases in tests we would need to add.

Reply:
> individual runtime.Objects - can we switch to something like RawSerializer#doEncode that is more robust?

Do encode doesn't write the SizeOfVarint before writing object. This cannot be easily fixed as it requires shifting bytes.

Reply:
Done, thanks for PoC.